In [1]:
# !pip install zipfile
# !pip install kaggle
# !pip install pandas

In [2]:
# Import the libraries needed for the project execution.
import pandas as pd
import zipfile
import kaggle

How to use kaggle API-> https://github.com/Kaggle/kaggle-api or https://www.kaggle.com/docs/api

In [3]:
# Download the dataset from kaggle using the Kaggle api
# The downloaded files will be at where this python file is running from
!kaggle datasets download -d hmavrodiev/london-bike-sharing-dataset


  0%|          | 0.00/165k [00:00<?, ?B/s]
100%|##########| 165k/165k [00:00<00:00, 195kB/s]
100%|##########| 165k/165k [00:00<00:00, 194kB/s]


In [4]:
# Extract the file from the downloaded zip file
zipfile_name = "london-bike-sharing-dataset.zip" # press tab between "" to find the name of the zip file
with zipfile.ZipFile(zipfile_name,'r') as file:
    file.extractall()
    
# Now a new file named "london-merged.csv" has been added to the folder

In [5]:
# Read the csv file using the pandas library as a dataframe
bikes = pd.read_csv("london_merged.csv")

In [6]:
# Look at the data we are working with
bikes.head()

timestamp  cnt   t1   t2    hum  wind_speed  weather_code  \
0  2015-01-04 00:00:00  182  3.0  2.0   93.0         6.0           3.0   
1  2015-01-04 01:00:00  138  3.0  2.5   93.0         5.0           1.0   
2  2015-01-04 02:00:00  134  2.5  2.5   96.5         0.0           1.0   
3  2015-01-04 03:00:00   72  2.0  2.0  100.0         0.0           1.0   
4  2015-01-04 04:00:00   47  2.0  0.0   93.0         6.5           1.0   

   is_holiday  is_weekend  season  
0         0.0         1.0     3.0  
1         0.0         1.0     3.0  
2         0.0         1.0     3.0  
3         0.0         1.0     3.0  
4         0.0         1.0     3.0

In [7]:
# Explore the data
bikes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17414 entries, 0 to 17413
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   timestamp     17414 non-null  object 
 1   cnt           17414 non-null  int64  
 2   t1            17414 non-null  float64
 3   t2            17414 non-null  float64
 4   hum           17414 non-null  float64
 5   wind_speed    17414 non-null  float64
 6   weather_code  17414 non-null  float64
 7   is_holiday    17414 non-null  float64
 8   is_weekend    17414 non-null  float64
 9   season        17414 non-null  float64
dtypes: float64(8), int64(1), object(1)
memory usage: 1.3+ MB


In [8]:
bikes.shape

(17414, 10)

So, as seen in the output above, the data has 10 columns and 17,414 rows of data

In [9]:
# See important information about that data and columns like their central tendency and their variability
bikes.describe()

cnt            t1            t2           hum    wind_speed  \
count  17414.000000  17414.000000  17414.000000  17414.000000  17414.000000   
mean    1143.101642     12.468091     11.520836     72.324954     15.913063   
std     1085.108068      5.571818      6.615145     14.313186      7.894570   
min        0.000000     -1.500000     -6.000000     20.500000      0.000000   
25%      257.000000      8.000000      6.000000     63.000000     10.000000   
50%      844.000000     12.500000     12.500000     74.500000     15.000000   
75%     1671.750000     16.000000     16.000000     83.000000     20.500000   
max     7860.000000     34.000000     34.000000    100.000000     56.500000   

       weather_code    is_holiday    is_weekend        season  
count  17414.000000  17414.000000  17414.000000  17414.000000  
mean       2.722752      0.022051      0.285403      1.492075  
std        2.341163      0.146854      0.451619      1.118911  
min        1.000000      0.000000      0.000000      0.000000  
25%        1.000000      0.000000      0.000000      0.000000  
50%        2.000000      0.000000      0.000000      1.000000  
75%        3.000000      0.000000      1.000000      2.000000  
max       26.000000      1.000000      1.000000      3.000000

In [10]:
# Count the unique values present in the weather_code feature/column
bikes['weather_code'].value_counts()

1.0     6150
2.0     4034
3.0     3551
7.0     2141
4.0     1464
26.0      60
10.0      14
Name: weather_code, dtype: int64

In [11]:
# Count the unique values present in the season feature/column
bikes['season'].value_counts()

0.0    4394
1.0    4387
3.0    4330
2.0    4303
Name: season, dtype: int64

So, the data is quite balanced in regards with the season column. The data is not balanced when looked at through weather code but unlike seasons weather is not something that is consistent in the first place

In [12]:
bikes.columns

Index(['timestamp', 'cnt', 't1', 't2', 'hum', 'wind_speed', 'weather_code',
       'is_holiday', 'is_weekend', 'season'],
      dtype='object')

In [13]:
# Specifying new columns names that will be easy to understand
new_cols = {'timestamp':'time',
            'cnt':'count',
            't1':'real_temp_C',
            't2':'temp_felt_like_C',
            'hum':'humidity_percentage',
            'wind_speed':'wind_speed_kph',
            'weather_code':'weather',
            'is_holiday':'is_holiday',
            'is_weekend':'is_weekend',
            'season':'season'}

# Renaming the columns to the specified names
bikes.rename(new_cols,axis=1,inplace=True)
bikes.head()

time  count  real_temp_C  temp_felt_like_C  \
0  2015-01-04 00:00:00    182          3.0               2.0   
1  2015-01-04 01:00:00    138          3.0               2.5   
2  2015-01-04 02:00:00    134          2.5               2.5   
3  2015-01-04 03:00:00     72          2.0               2.0   
4  2015-01-04 04:00:00     47          2.0               0.0   

   humidity_percentage  wind_speed_kph  weather  is_holiday  is_weekend  \
0                 93.0             6.0      3.0         0.0         1.0   
1                 93.0             5.0      1.0         0.0         1.0   
2                 96.5             0.0      1.0         0.0         1.0   
3                100.0             0.0      1.0         0.0         1.0   
4                 93.0             6.5      1.0         0.0         1.0   

   season  
0     3.0  
1     3.0  
2     3.0  
3     3.0  
4     3.0

In [14]:
# Changing the humidity values to percentages (i.e., a value between 0 and 1)
bikes['humidity_percentage'] = bikes['humidity_percentage']/100

In [15]:
# Createing a season dictionary so that the values 0 to 3 can be mapped to the real world terms
season_dict = {
    0.0 : "Spring",
    1.0 : "Summer",
    2.0 : "Autumn",
    3.0 : "Winter"
}

# Creating a weather dictionary so that the values like 1,2,3,10 etc, can be mapped to the real world terms
weather_dict = {
    1.0 : "Clear",
    2.0 : "Scattered Clouds",
    3.0 : "Broken Clouds",
    4.0 : "Cloudy",
    7.0 : "Rain",
    10.0 : "Rain with Thunderstorms",
    26.0 : "Snowfall"
}

# Mapping the values 0-3 to the actual written seasons
bikes['season'] = bikes['season'].map(season_dict)

# Mapping the weather code values to the actual weather conditions
bikes['weather'] = bikes['weather'].map(weather_dict)

In [16]:
# Checking the dataframe to see whether the mappings have worked
bikes.head()

time  count  real_temp_C  temp_felt_like_C  \
0  2015-01-04 00:00:00    182          3.0               2.0   
1  2015-01-04 01:00:00    138          3.0               2.5   
2  2015-01-04 02:00:00    134          2.5               2.5   
3  2015-01-04 03:00:00     72          2.0               2.0   
4  2015-01-04 04:00:00     47          2.0               0.0   

   humidity_percentage  wind_speed_kph        weather  is_holiday  is_weekend  \
0                0.930             6.0  Broken Clouds         0.0         1.0   
1                0.930             5.0          Clear         0.0         1.0   
2                0.965             0.0          Clear         0.0         1.0   
3                1.000             0.0          Clear         0.0         1.0   
4                0.930             6.5          Clear         0.0         1.0   

   season  
0  Winter  
1  Winter  
2  Winter  
3  Winter  
4  Winter

In [17]:
# Writing the final dataframe to an excel file that we can then use in Tableau visualization.
# The file name will be "London Bike Sharing Dataset.xlsx"
bikes.to_excel("London Bike Sharing Dataset.xlsx",sheet_name="Data",index=False)

# Done!!